File: client/src/services/gifs.ts
import { apiFetch } from './api';
import type { Media } from './media';

export type KlipyKind = 'gifs' | 'stickers' | 'memes';

export interface GifSearchItem {
  id: string;
  slug: string;
  title: string;
  preview_url: string;
  gif_url: string;
  width: number;
  height: number;
  kind?: string;
}

export interface GifSearchResponse {
  enabled: boolean;
  items: GifSearchItem[];
  page: number;
  per_page: number;
  has_next: boolean;
  attribution: string;
  kind?: string;
}

/** Proxy → KLIPY search/trending. Empty `q` returns trending. */
export async function searchGifs(params?: {
  q?: string;
  kind?: KlipyKind;
  page?: number;
  perPage?: number;
}): Promise<GifSearchResponse> {
  const qs = new URLSearchParams();
  if (params?.q?.trim()) qs.set('q', params.q.trim());
  qs.set('kind', params?.kind ?? 'gifs');
  if (params?.page) qs.set('page', String(params.page));
  if (params?.perPage) qs.set('per_page', String(params.perPage));
  return apiFetch<GifSearchResponse>(`/v1/gifs/search?${qs.toString()}`);
}

/** Download a KLIPY asset into the ScreenRaid media library. */
export async function importGif(params: {
  url: string;
  title?: string;
  slug?: string;
  roomId?: string;
  kind?: KlipyKind;
}): Promise<Media> {
  return apiFetch<Media>('/v1/gifs/import', {
    method: 'POST',
    body: JSON.stringify({
      url: params.url,
      title: params.title ?? null,
      slug: params.slug ?? null,
      room_id: params.roomId ?? null,
      kind: params.kind ?? 'gifs',
    }),
  });
}

File: server/src/api/handlers/gifs.rs
//! Authenticated KLIPY search / import endpoints (gifs, stickers, memes).

use axum::{
    extract::{Query, State},
    http::StatusCode,
    Json,
};
use serde::Deserialize;
use uuid::Uuid;

use crate::api::middleware::auth::AuthUser;
use crate::error::AppError;
use crate::service::gif_service::GifSearchResponse;
use crate::state::AppState;

#[derive(Debug, Deserialize)]
pub struct GifSearchQuery {
    pub q: Option<String>,
    /// `gifs` | `stickers` | `memes`
    #[serde(default = "default_kind")]
    pub kind: String,
    #[serde(default = "default_page")]
    pub page: u32,
    #[serde(default = "default_per_page")]
    pub per_page: u32,
}

fn default_kind() -> String {
    "gifs".into()
}

fn default_page() -> u32 {
    1
}

fn default_per_page() -> u32 {
    24
}

#[derive(Debug, Deserialize)]
pub struct ImportGifRequest {
    pub url: String,
    pub title: Option<String>,
    pub slug: Option<String>,
    pub room_id: Option<Uuid>,
    pub kind: Option<String>,
}

pub async fn search_gifs(
    _auth: AuthUser,
    State(state): State<AppState>,
    Query(query): Query<GifSearchQuery>,
) -> Result<Json<GifSearchResponse>, AppError> {
    Ok(Json(
        state
            .gifs
            .search(
                &query.kind,
                query.q.as_deref(),
                query.page,
                query.per_page,
            )
            .await?,
    ))
}

pub async fn import_gif(
    auth: AuthUser,
    State(state): State<AppState>,
    Json(body): Json<ImportGifRequest>,
) -> Result<(StatusCode, Json<screenraid_types::Media>), AppError> {
    let media = state
        .gifs
        .import_url(
            auth.user_id,
            body.room_id,
            &body.url,
            body.title.as_deref(),
            body.slug.as_deref(),
            body.kind.as_deref(),
        )
        .await?;
    Ok((StatusCode::CREATED, Json(media)))
}
